pennylane/transforms/compile.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Code for a compilation transform."""
+
+from pennylane.transforms import qfunc_transform, cancel_inverses, cnot_to_cz
+
+simple_pipeline = [cnot_to_cz, cancel_inverses]
+
+
+@qfunc_transform
+def compile(tape, pipeline=simple_pipeline, basis_set=None, num_passes=1):
+    """Compile a circuit by applying a series of transforms to a quantum function.
+
+    Args:
+        tape (.QuantumTape): A quantum tape.
+        pipeline (List[qfunc_transform]): A list of quantum function transforms to apply.
+        basis_set (List[str]): A list of basis gates. When expanding the tape, expansion
+            will continue until gates in the specific set are reached. If no basis set is
+            specified, no expansion will be done.
+        num_passes (int): The number of times to repeat the set of operations in pipeline.
+
+    **Example**
+
+    Consider the following quantum function:
+
+    .. code-block:: python
+
+        def bell_state():
+            qml.Hadamard(wires=0)
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+    Suppose we would like to apply the following set of transforms:
+
+    >>> my_transforms = [cnot_to_cz, cancel_inverses]
+
+    We can use this as a pipeline for the ``compile`` transform:
+
+    >>> compiled_bell_state = compile(pipeline=my_transforms)(bell_state)
+
+    Now we can run this, and the compiled version will be executed:
+
+    >>> dev = qml.device('default.qubit', wires=2)
+    >>> qnode = qml.QNode(compiled_bell_state, dev)
+    >>> print(qml.draw(qnode)())
```

Comment:
A big disadvantage here is that `compile` is run every time the QNode is executed. I think it's pretty critical to find a way to cache the results somehow.

Reply:
💯 

I think this will actually work out-of-the-box if you do

```python
@qml.qnode(dev, mutable=False)
```

However, there is a subtlety --- this assume that the circuit structure is *independent* of the QNode parameters. That is, you can't have something like
```python
for p in weights:
    qml.RX(p, wires=0)
```
and take advantage of the fact that calling the qnode with different length `weights` input will result in different circuits.

If the QNode circuit structure _is_ a function of the QNode parameters, this is a much harder problem. We are looking at something more along the lines of a JIT --- that is, compile the quantum function down to a function in a higher-performance low-level language, and use _that_ compiled function on future calls.

Reply:
Oh gosh there is a huge bug in mutable mode 😆 It's still using the parameter values from the first execution. Don't use it for now lol

Reply:
Have just tested by throwing a print in `compile`; I noticed that with `mutable=False`, executing the QNode multiple times doesn't call `compile` again, however if I then do `qml.draw()`, `compile` is executed again. Is that expected behaviour? (Like if the QNode itself goes into a transform, the tape is remade?)

Reply:
> Oh gosh there is a huge bug in mutable mode laughing It's still using the parameter values from the first execution. Don't use it for now lol

:grimacing: Bug notwithstanding, I think this is a pretty good solution. If you have a circuit structure that depends on parameters, you would need to recompile it anyways if the structure changes. (Or, is it the case that, even if the structure could change but doesn't, like if you just want to change the values of input parameters but not the amount of them, it would *still* recompile?)

Reply:
> (Or, is it the case that, even if the structure could change but doesn't, like if you just want to change the values of input parameters but not the amount of them, it would still recompile?)

I think so yes. We need to invest it building out a `mutable` option that is 'smarter'.